lib/elixir/lib/inspect/algebra.ex

```diff
@@ -478,26 +478,33 @@ defmodule Inspect.Algebra do
     "..."
   end
 
-  defp do_surround_many([h], limit, opts, fun, _sep) do
+  defp do_surround_many([], _limit, _opts, _fun, _sep) do
+    :doc_nil
+  end
+
+  defp do_surround_many([h], limit, opts, fun, sep) do
     fun.(h, %{opts | limit: limit})
   end
 
   defp do_surround_many([h|t], limit, opts, fun, sep) when is_list(t) do
     limit = decrement(limit)
-    glue(
-      concat(fun.(h, %{opts | limit: limit}), sep),
-      do_surround_many(t, limit, opts, fun, sep)
-    )
+    h = fun.(h, %{opts | limit: limit})
+    t = do_surround_many(t, limit, opts, fun, sep)
+    do_join(h, t, sep)
   end
 
-  defp do_surround_many([h|t], limit, opts, fun, _sep) do
+  defp do_surround_many([h|t], limit, opts, fun, sep) do
```

Comment:
So this line is being called when we have an improper list. And we don't have tests for those. :)